atomix/cluster/src/test/java/io/atomix/raft/ControllableRaftContexts.java

```diff
@@ -344,4 +360,66 @@ private void updateAndVerifyLeaderTerm(final RaftContext s) {
       }
     }
   }
+
+  // If a node is in CandidateRole, then it will update the term. But until the election is
+  // completed, there is no leader at that term.
+  public boolean hasLeaderAtTheLatestTerm() {
+    final var currentTerm =
+        raftServers.values().stream().map(RaftContext::getTerm).max(Long::compareTo).orElseThrow();
+    return leadersAtTerms.get(currentTerm) != null;
+  }
+
+  boolean hasCommittedAllEntries() {
+    return raftServers.values().stream()
+        .noneMatch(
+            s -> {
+              final var lastCommittedEntry = getLastCommittedEntry(s);
+              final var lastUncommittedEntry = getLastUncommittedEntry(s);
+
+              return lastUncommittedEntry != null
+                  && !lastUncommittedEntry.equals(lastCommittedEntry);
+            });
+  }
```

Comment:
:wrench: I found this method to be a bit difficult to understand, maybe we could simplify?

```java
boolean hasCommittedAllEntries() {
  return raftServers.values().stream().allMatch(s -> getLastUncommittedEntry(s) == getLastCommittedEntry(s));
}
```

Reply:
I think `==` doesn't work here because each reader creates a different object for the same entry. But I will use `allMatch` instead of `noneMatch`.